CMakeLists.txt

```diff
@@ -0,0 +1,225 @@
+cmake_minimum_required(VERSION 3.14)
+project(mold VERSION 1.3.0)
+
+include(CMakeDependentOption)
+include(GNUInstallDirs)
+
+# Mold build options
+cmake_dependent_option(MOLD_USE_MIMALLOC "Use mimalloc" ON
+    "NOT CMAKE_SYSTEM_NAME MATCHES Darwin; NOT Android" OFF)
+option(MOLD_USE_SYSTEM_MIMALLOC "Use system mimalloc")
+option(MOLD_USE_SYSTEM_TBB "Use system TBB")
+option(MOLD_USE_LTO "Use Link Time Optimization")
+option(CMAKE_INSTALL_LOCAL_ONLY ON)
+
+# Default to `Release` build type,
+# to be consistent with the current Makefiles
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE Release CACHE STRING "" FORCE)
+endif()
+
+# Setup mimalloc
+if(MOLD_USE_MIMALLOC)
+    if(MOLD_USE_SYSTEM_MIMALLOC)
+        find_library(MIMALLOC_LIBRARIES mimalloc REQUIRED)
+        add_library(mimalloc INTERFACE)
+        target_link_libraries(mimalloc INTERFACE ${MIMALLOC_LIBRARIES})
+        target_compile_definitions(mimalloc INTERFACE USE_SYSTEM_MIMALLOC)
+    else()
+        function(mold_add_mimalloc)
+            string(APPEND CMAKE_C_FLAGS -DMI_USE_ENVIRON=0)
+            option(MI_BUILD_TESTS "Build test executables" OFF)
+            add_subdirectory(third-party/mimalloc EXCLUDE_FROM_ALL)
+        endfunction()
+
+        mold_add_mimalloc()
+    endif()
+
+    add_library(mimalloc::mimalloc ALIAS mimalloc)
```

Comment:
I know changing the `CMAKE_C_FLAGS` is bad. But I just tried to port the existing Makefile as close as possible.
Also, what this the advantage of this `Findmimalloc.cmake` script?

Reply:
mimalloc already supports clients using CMake, because the project exports a CMake package. That should be used preferably. The approach with the find module makes vendoring opt-in, but the project code just has the idiomatic `find_package(mimalloc REQUIRED)` call that one would expect regardless. The linked example project illustrates this.

Interesting, I opened the github.dev page for mimalloc and project wide search didn't return anything at first for that macro. In this case you can just use `add_compile_definitions(MI_USE_ENVIRON=0)` instead.

Reply:
mhm `add_compile_definitions` does the the define for the current directory and below, so it also sets this for the mold target...